src/e2e_tests.rs

```diff
@@ -0,0 +1,649 @@
+            .context("unable to get response")?;
+        let mut response: Message = serde_json::from_str::<Message>(&result.unwrap().1)
+            .context("unable to parse response")?;
+
+        (response.destination, response.source) = (vec![response.source], response.destination[0]);
+        let _ = conn
+            .lpush(&response.reply, &response)
+            .await
+            .context("unable to push response")?;
+    }
+}
+
+async fn start_rmb<
+    D: TwinDB + Clone,
+    S: Storage + ProxyStorage + Clone,
+    I: Identity + Signer + 'static,
+>(
+    db: D,
+    storage: S,
+    ident: I,
+    address: &str,
+    id: u32,
+) -> Result<()> {
+    let processor_handler = tokio::spawn(processor(id, storage.clone()));
+
+    let api_handler =
+        tokio::spawn(HttpApi::new(id, address, storage.clone(), ident.clone(), db.clone())?.run());
+
+    let proxy_handler = tokio::spawn(ProxyWorker::new(id, 10, storage.clone(), db.clone()).run());
+
+    let workers_handler = tokio::task::spawn(HttpWorker::new(1000, storage, db, ident).run());
+
+    tokio::select! {
+        result = processor_handler => {
+            if let Err(err) = result {
+                bail!("message processor panicked unexpectedly: {}", err);
+            }
+        },
+        result = api_handler => {
+            match result {
+                Err(err) => bail!("http server panicked unexpectedly: {}", err),
+                Ok(Ok(_)) => bail!("http server exited unexpectedly"),
+                Ok(Err(err)) => bail!("http server exited with error: {}", err),
+            }
+        },
+        result = workers_handler => {
+            if let Err(err) = result {
+                bail!("http workers panicked unexpectedly: {}", err);
+            }
+        },
+        result = proxy_handler => {
+            if let Err(err) = result {
+                bail!("proxy workers panicked unexpectedly: {}", err);
+            }
+        },
+    };
+    unreachable!();
+}
+enum KPair {
+    EdPair,
+    SrPair,
+}
+
+fn create_test_signer(p: KPair) -> Result<impl Signer> {
+    match p {
+        KPair::EdPair => {
+            let (_pair, string, _seed) = EdPair::generate_with_phrase(None);
+            let ed_signer = identity::Ed25519Signer::try_from(string.as_ref()).unwrap();
+            Ok(ed_signer)
+        }
+        KPair::SrPair => {
+            let (_pair, string, _seed) = SrPair::generate_with_phrase(None);
+            let sr_signer = identity::Ed25519Signer::try_from(string.as_ref()).unwrap();
+            Ok(sr_signer)
+        }
+    }
+}
+
+async fn create_local_redis_storage(port: usize) -> Result<impl Storage + ProxyStorage> {
+    let redis_url = format!("redis://localhost:{}", port);
+    let pool = redis::pool(redis_url)
+        .await
+        .context("unable to create redis connection")?;
+    let storage = RedisStorage::builder(pool).build();
+    Ok(storage)
+}
+
+fn create_mock_db(twins: Vec<&Twin>) -> Result<impl TwinDB + Clone> {
```

Comment:
There is no actual benefit in this case to return `Result<impl TwinDB + Clone>` because there is no dynamic allocations `Box` or anything else. So just return `Result<InMemoryDB>`